cmd/imagec/docker.go

```diff
@@ -269,9 +257,35 @@ func FetchImageBlob(options ImageCOptions, image *ImageWithMeta) (string, error)
 
 	diffID = fmt.Sprintf("sha256:%x", diffIDSum.Sum(nil))
 
+	// this isn't an empty layer, so we need to calculate the size
 	if diffID != string(DigestSHA256EmptyTar) {
-		// bytesRead represents the size of the decompressed layer data
-		image.size = bytesRead
+		var layerSize int64
+
+		// seek to the beginning of the file
+		imageFile.Seek(0, 0)
+
+		// recreate the decompressed tar Reader
+		decompressedTar, err := archive.DecompressStream(imageFile)
+		if err != nil {
+			return "", err
+		}
+
+		// get a tar reader for access to the files in the archive
+		tr := tar.NewReader(decompressedTar)
+
+		// iterate through tar headers to get file sizes
+		for {
+			tarHeader, err := tr.Next()
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				return "", err
+			}
+			layerSize += tarHeader.Size
+		}
+
+		image.size = layerSize
 	}
```

Comment:
Yeah but that's tar file itself. I believe tar has a header that contains the original file size (which you use above) but to reach the tar we need to decompress gz part, right? So if we download a 50MB tar.gz file, and if unpacking that ends up creating a 75MB tar file, then we will have 75MB in memory to iterate over its header. 

By the way I'm just trying to understand and not opposing or suggesting something, to be honest I don't even know whether it would be possible to get tar header without decompressing the whole gzip part.